### ISSUE TITLE
[ecobee] Unable to complete API call because token is expired - No new token

### ISSUE THREAD

**Issue by morph166955** [openhab/openhab-addons#9039]
My ecobee api connection abruptly stopped this morning with this error...

```
2020-11-15 10:05:32.196 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 10:05:32.196 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
```

It looks like this was added in https://github.com/openhab/openhab-addons/pull/7506

I have not found a way to actually refresh the token.  No new PIN has been presented to me.

**Comment by morph166955** [comment]
Workaround:

Shut off openhab
remove oauth info out of /var/lib/openhab/jsondb/
remove app from ecobee website
restart openhab 

This triggered the 4 character pin request.

**Comment by mhilbush** [comment]
Did you get this warn in the log in addition to the log entries you showed above?

https://github.com/mhilbush/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java#L327

If not, I don't quite understand why the communication would've stopped. The absence of that warn log entry suggests the second try to get the access token was successful.

**Comment by morph166955** [comment]
I do not.  This is 10 minutes of errors:

```
 2020-11-15 09:50:19.027 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:50:19.027 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:50:22.092 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:50:22.092 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:50:45.679 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:50:45.680 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:50:54.115 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:50:54.115 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:51:25.513 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:51:25.513 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:51:57.271 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:51:57.271 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:52:19.000 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:52:19.000 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:52:28.820 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:52:28.821 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:53:00.254 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:53:00.254 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:53:31.646 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:53:31.646 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:54:03.078 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:54:03.079 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:54:19.090 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:54:19.091 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:54:34.471 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:54:34.471 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:55:06.114 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:55:06.114 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:55:37.708 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:55:37.708 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:55:52.096 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:55:52.096 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:56:09.589 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:56:09.589 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:56:19.120 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:56:19.120 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:56:40.982 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:56:40.982 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:57:12.385 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:57:12.385 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:57:43.810 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:57:43.811 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:58:15.443 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:58:15.444 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:58:19.089 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:58:19.089 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:58:46.862 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:58:46.862 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:59:18.394 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:59:18.394 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired
2020-11-15 09:59:49.824 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2020-11-15 09:59:49.824 [INFO ] [inding.ecobee.internal.api.EcobeeApi] - API: Unable to complete API call because token is expired

```

**Comment by morph166955** [comment]
I can say that I was not getting any data from ecobee.  All items were null.

**Comment by mhilbush** [comment]
Would've been great to get a debug log, but I suppose that's not possible at this point?

**Comment by morph166955** [comment]
Unfortunately not, I killed the token and reset to get back functional.  I'll grab logs if this happens again.

**Comment by mhilbush** [comment]
Actually, I think I see a potential logic error. If [this call](https://github.com/mhilbush/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java#L326) to `isAuthorized`  returns true, I should not be setting `success` to false [here](https://github.com/mhilbush/openhab-addons/blob/2.5.x/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java#L330).

**Comment by morph166955** [comment]
One thing to note, the things were all online while this was happening.

**Comment by mhilbush** [comment]
@morph166955 I made a fix for the logic error. While I'm not convinced the fix will fully resolve the issue you were having, at least that situation will be handled correctly in the code. You're more than welcome to run that version. Just let me know and I'll post it for you to download.

Ultimately, I may need to talk with Ecobee support to better understand what's happening and how best to handle it.

By any chance, were you doing a lot of stops/starts of openHAB when the situation occurred. In reading through some posts of other smarthome apps, that seemed to be a trigger of that error.

**Comment by morph166955** [comment]
I was actually restarting several times to fix other issues.

**Comment by mhilbush** [comment]
Sounds good. If you want the jar, just ping me here or on the Ecobee forum thread. In the meantime, I'll probably submit the fix for inclusion in 3.0.

If you feel like tweaking your logging config, you could run the binding in DEBUG mode just in case it happens again. If you want, I can give you a logger config that puts the Ecobee logs in a separate file.

**Comment by mhilbush** [comment]
This is closed with #9109 and #9065
